apt/minimization/minimizer.py

```diff
@@ -315,30 +392,57 @@ def transform(self, X):
             # values (leaves others untouched)
             if indexes and not representatives.columns.empty:
                 if len(indexes) > 1:
-                    replace = pd.concat([representatives.loc[i].to_frame().T]*len(indexes)).reset_index(drop=True)
+                    replace = pd.concat([representatives.loc[i].to_frame().T] * len(indexes)).reset_index(drop=True)
                 else:
                     replace = representatives.loc[i].to_frame().T.reset_index(drop=True)
                 replace.index = indexes
                 generalized.loc[indexes, representatives.columns] = replace
-
-        return generalized.to_numpy()
+        if self.is_numpy:
+            return generalized.to_numpy()
+        return generalized
 
     def _get_record_indexes_for_cell(self, X, cell, mapped):
-        return [i for i, x in enumerate(X) if not mapped.item(i) and
-                self._cell_contains(cell, x, i, mapped)]
+        indexes = []
+        for index, row in X.iterrows():
+            if not mapped.item(index) and self._cell_contains(cell, row, index, mapped):
+                indexes.append(index)
+        return indexes
 
     def _cell_contains(self, cell, x, i, mapped):
         for f in self._features:
             if f in cell['ranges']:
                 if not self._cell_contains_numeric(f, cell['ranges'][f], x):
                     return False
+            elif f in cell['categories']:
+                if not self._cell_contains_categorical(f, cell['categories'][f], x):
+                    return False
+            elif f in cell['untouched']:
+                continue
             else:
-                #TODO: exception - feature not defined
-                pass
+                raise TypeError("feature not defined")
         # Mark as mapped
         mapped.itemset(i, 1)
         return True
 
+    def _modify_categorical_features(self, X):
+        self.categorical_values = {}
+        self.oneHotVectorFeaturesToFeatures = {}
+        features_to_remove = []
+        for feature in self.categorical_features:
+            try:
+                all_values = X.loc[:, feature]
+                values = list(all_values.unique())
+                self.categorical_values[feature] = values
+                X[feature] = pd.Categorical(X.loc[:, feature], categories=values, ordered=False)
+                ohe = pd.get_dummies(X[feature], prefix=feature)
+                for oneHotVectorFeature in ohe.columns:
+                    self.oneHotVectorFeaturesToFeatures[oneHotVectorFeature] = feature
+                X = pd.concat([X, ohe], axis=1)
+                features_to_remove.append(feature)
+            except KeyError:
+                print('feature not found: ' + feature)
```

Comment:
Let's also please make this print more clear, like what feature was not found and where.